src/nodes.coffee

```diff
@@ -266,43 +270,99 @@ exports.Base = class Base
-  # as JSON. This is used for generating an abstract syntax tree (AST).
-  # This is what the `ast` option in the Node API returns.
-  toJSON: ->
-    # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babylon/ast/spec.md)
-    # as closely as possible, for improved interoperability with other tools.
-    obj =
-      type: @constructor.name
-      # Convert `locationData` to Babel’s style.
-      loc:
-        start:
-          line: @locationData.first_line
-          column: @locationData.first_column
-        end:
-          line: @locationData.last_line
-          column: @locationData.last_column
-
-    # Add serializable properties to the output. Properties that aren’t
-    # automatically serializable (because they’re already a primitive type)
-    # should be handled on a case-by-case basis in child node classes’ own
-    # `toJSON` methods.
-    for property, value of this
-      continue if property in ['locationData', 'children']
-      continue if value is undefined # Don’t skip `null` or `false` values.
-      if typeof value is 'boolean' or typeof value is 'number' or typeof value is 'string'
-        obj[property] = value
-
-    # Work our way down the tree. This is like `eachChild`, except that we
-    # preserve the child node name, and arrays.
-    for attr in @children when @[attr]
-      if Array.isArray(@[attr])
-        obj[attr] = []
-        for child in flatten [@[attr]]
-          obj[attr].push child.unwrap().toJSON()
-      else
-        obj[attr] = @[attr].unwrap().toJSON()
+  # as JSON. This is what the `ast` option in the Node API returns.
+  # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babylon/ast/spec.md)
+  # as closely as possible, for improved interoperability with other tools.
+  toAst: (o, level) ->
+    o = extend {}, o
+    o.level = level if level
+    @withAstLocationData @withAstType @getAstContent(o), o
+
+  # By default, a node class's AST `type` is the class name
+  astType: -> @constructor.name
+
+  # Adds `type` to an AST.
+  # A node class typically defines `astType` (as a string or callback) if it
+  # needs to override the default-generated `type` (ie its constructor name)
+  withAstType: (ast, o) ->
+    return ast unless ast
+    return ast if Array.isArray ast
+    return ast if ast.type
+    return ast unless @emptyAst or do ->
+      return yes for key in Object.keys(ast) when key not in ['comments', astLocationFields...]
+
+    merge ast,
+      type: @astType?(o) ? @astType
+
+  # Returns the "content" (ie not `type` or location data) of the AST for the node.
+  # By default, recursively generates AST nodes for `children`.
+  # To override, a node class can either:
+  #   - define `astChildren` to specify how to generate full child nodes, and/or
+  #     `astProps` to specify how to generate non-child-node fields
+  #   - override the `getAstContent()` method
+  getAstContent: (o) ->
```

Comment:
> we need a new name for `_toAst`, not least because we don’t have `_`-prefixed methods anywhere else in the codebase

@GeoffreyBooth ok that makes sense, updated `_toAst()` -> `getAstContent()`

> What is that method doing that’s different than `toAst`?

Hopefully `getAstContent()` is a more descriptive name that helps explain what its role is, it's expected to return all of the "content fields" of the AST representation (ie everything except `type` and location data)

> Can it be split up?

In a sense it is "split up" (if you look at its base-class implementation just below) into the "child nodes" (overridable via `astChildren`) and "simple fields" (overridable via `astProps`), which I think is a valuable conceptual distinction to make and allows `astChildren` to have the "smart default" of `children` (independently of specifying custom simple fields via `astProps`)

> It still feels wrong to me that we have three methods here (`toJSON`, `toAst`, `_toAst`)

Ok so what I did is to remove `toJSON()`. I'm being slightly provocative, but see eg [this reply](https://github.com/jashkenas/coffeescript/pull/5099#discussion_r212808123) as far as why I think it's questionable whether the original thinking behind using the name `toJSON()` even applies. And I feel pretty strongly that `toJSON()` is a bad, obfuscating name to be using in various places around `nodes.coffee` (if I were to have renamed `toAst()` -> `toJSON()`)

So now there's just `toAst()` (which is what `coffeescript.coffee` now calls as the entry point for AST generation) and `getAstChildren()`. If we can try it like this, I'm guessing you'll have a chance to see and grasp the usage patterns (within `nodes.coffee`) over the course of reviewing indivual-node-type-AST PRs. If by the end of that (ie when we're close to ready to actually merge `ast` into `master`) you or someone else feels strongly about `toJSON()` being used, at that point we could easily either rename `toAst()` -> `toJSON()` or perhaps my pattern of "just use `toJSON()` as an entry-point name and then immediately hand off to the internal `toAst()`" would be clearer at that point in light of the way things are actually being used

> or perhaps better would be `toAstJSON`

I don't think that's a better name than `toAst()`. It's not technically returning JSON, everything that uses its return value inside `nodes.coffee` is just engaging with it as a normal JS object. And I don't think (if there's a need at all) that this name is the best place to alert people to the "format" of the AST we're producing (ie that it's a JS object/could be converted to JSON)

> Perhaps `compileNode` should become `toJavaScriptString`, and the various `compile*` methods renamed accordingly (`toJavaScriptStringFragments`, etc.)

Interesting thought, but see my other reply just now about why I think doing unnecessary renaming/refactoring of existing JS generation code is a bad idea at least to include in this scope

> That would also make it obvious which methods to remove once we have the printer up and running

I don't think that's valid, they'll still be obvious with their current names

> Sorry if it seems like I’m splitting hairs on bikeshed stuff, but I’m trying to understand the overall architecture of what you’re doing

Ya it's obviously valuable for you to understand it to the point of being comfortable using/manipulating it. And naming is important. I'd just say that this isn't our only opportunity to make decisions about this or for you to wrap your head around it - I think that starting to review individual-node-type-AST PRs will give you a lot of insight into the usage patterns and hopefully the "why" of some of the decisions I've made, and we can always continue these discussions/revisit some of the decisions as we're going through those subsequent PRs